File: backend/simple_expr.py
import re
import typing


class ExpressionError(Exception):
    pass


def normalize_whitespace(s: str) -> str:
    return re.sub(r'\s+', ' ', s)


def matcher_word(name: str):
    return rf'(?P<{name}>\w+)'


def matcher_int(name: str):
    return rf'(?P<{name}>[0-9]+)'


def matcher_number(name: str, decimal_symbol: str = 'dot'):
    if decimal_symbol == 'dot':
        return rf'(?P<{name}>[0-9 ,]+(\.[0-9]*)?)' # TODO better handling of leading/trailing whitespace
    elif decimal_symbol == 'comma':
        return rf'(?P<{name}>[0-9 .]+(,[0-9]*)?)' # TODO better handling of leading/trailing whitespace
    else:
        raise ExpressionError('number type must be "dot" or "comma"')


SIMPLE_EXPR_TO_REGEX: dict[str, typing.Callable[[str], str]] = {
    'word': matcher_word,
    'int': matcher_int,
    'integer': matcher_int,
    'number': matcher_number,
    'decimal': matcher_number
}


def simple_expr_to_regex(expr: str) -> str:
    # Split on < or >, unless preceded by a backslash. Include separators in result.
    parts = re.split(r'((?<!\\)[<>*?])', expr)

    regex = ''
    in_brackets = False
    while parts:
        word = parts.pop(0)
        if word == '<':
            in_brackets = True
            continue
        elif word == '>':
            in_brackets = False
            continue
        elif word == '*':
            regex += '.*?'
        elif word == '?':
            regex += '.'
        else:
            if in_brackets:
                placeholder_parts = word.strip().split(':')
                if len(placeholder_parts) < 2:
                    raise ExpressionError(f'Placeholder "{word}" must specify name and type')
                
                placeholder_name = placeholder_parts[0]
                placeholder_type = placeholder_parts[1].lower()
                placeholder_args = placeholder_parts[2:]

                if placeholder_name == '':
                    raise ExpressionError(f'Placeholder name can not be empty')

                if not re.match(r'^[a-zA-Z]\w*$', placeholder_name): # TODO allow unicode
                    raise ExpressionError(f'Placeholder name can not contain special characters')

                if placeholder_type not in SIMPLE_EXPR_TO_REGEX:
                    raise ExpressionError(f'Unknown placeholder type "{placeholder_type}"')
                
                as_re = SIMPLE_EXPR_TO_REGEX[placeholder_type](placeholder_name, *placeholder_args)
                regex += as_re

            else:
                regex += re.escape(normalize_whitespace(word)).replace(r'\ ', r'\s+')

    if in_brackets:
        raise ExpressionError('Placeholder not terminated')
    return regex

File: plngx-dissect-frontend/src/utils/simple_expr.ts
// Converted from Python to Typescript with Chat-GPT

// ExpressionError
export class ExpressionError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "ExpressionError";
  }
}

// normalize_whitespace
export function normalizeWhitespace(s: string): string {
  return s.replace(/\s+/g, " ");
}

// matchers
export function matcherWord(name: string): string {
  return `(?<${name}>\\p{Letter}+)`;
}

export function matcherInt(name: string): string {
  return `(?<${name}>[0-9]+)`;
}

export function matcherNumber(name: string, decimalSymbol: "dot" | "comma" = "dot"): string {
  if (decimalSymbol === "dot") {
    return `(?<${name}>[0-9 ,]+(\\.[0-9]*)?)`;
  } else if (decimalSymbol === "comma") {
    return `(?<${name}>[0-9 .]+(,[0-9]*)?)`;
  } else {
    throw new ExpressionError('number type must be "dot" or "comma"');
  }
}

// Mapping of simple expression types to regex generator functions
type MatcherFunc = (name: string, ...args: string[]) => string;

const SIMPLE_EXPR_TO_REGEX: Record<string, MatcherFunc> = {
  'word': matcherWord,
  'int': matcherInt,
  'integer': matcherInt,
  'number': matcherNumber,
  'decimal': matcherNumber,
};

// Convert a simple expression into a regex string
export function simpleExprToRegex(expr: string): string {
  // Split on < or >, unless preceded by a backslash. Include separators in result.
  const parts = expr.split(/((?<!\\)[<>*?])/);

  let regex = "";
  let inBrackets = false;

  while (parts.length > 0) {
    const word = parts.shift()!;
    if (word === "<") {
      inBrackets = true;
      continue;
    } else if (word === ">") {
      inBrackets = false;
      continue;
    } else if (word === "*") {
      regex += ".*?";
    } else if (word === '?') {
      regex += ".";
    } else {
      if (inBrackets) {
        const placeholderParts = word.trim().split(":");
        if (placeholderParts.length < 2) {
          throw new ExpressionError(`Placeholder "${word}" must specify name and type`);
        }

        const placeholderName = placeholderParts[0];
        const placeholderType = placeholderParts[1].toLowerCase();
        const placeholderArgs = placeholderParts.slice(2);

        if (!/^[a-zA-Z]\w*$/.test(placeholderName)) { // TODO allow unicode
          throw new ExpressionError(`Placeholder name can not contain special characters`);
        }

        if (!(placeholderType in SIMPLE_EXPR_TO_REGEX)) {
          throw new ExpressionError(`Unknown placeholder type "${placeholderType}"`);
        }

        const asRe = SIMPLE_EXPR_TO_REGEX[placeholderType](placeholderName, ...placeholderArgs);
        regex += asRe;
      } else {
        regex += RegExp.escape(normalizeWhitespace(word)).replace(/( |\\x20)/g, "\\s+");
      }
    }
  }
  if (inBrackets) {
    throw new ExpressionError('Placeholder not terminated');
  }

  return regex;
}
